src/drivers/uavcan/node_info.cpp

```diff
@@ -0,0 +1,335 @@
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#include "node_info.hpp"
+
+#include <px4_platform_common/log.h>
+#include <px4_platform_common/px4_config.h>
+#include <drivers/drv_hrt.h>
+#include <cstring>
+
+using namespace time_literals;
+
+NodeInfoPublisher::NodeInfoPublisher(uavcan::INode &node, uavcan::NodeInfoRetriever &node_info_retriever)
+	: TimerBase(node), _node_info_retriever(node_info_retriever)
+{
+	_node_info_retriever.addListener(this);
+}
+
+NodeInfoPublisher::~NodeInfoPublisher()
+{
+	_node_info_retriever.removeListener(this);
+	delete[] _device_informations;
+}
+
+void NodeInfoPublisher::handleNodeInfoRetrieved(uavcan::NodeID node_id,
+		const uavcan::protocol::GetNodeInfo_::Response &node_info)
+{
+	const NodeInfo info(node_id, node_info);
+	registerDevice(info.node_id.get(), &info, 0, DeviceCapability::NONE);
+
+	startTimerIfNotRunning();
+}
+
+void NodeInfoPublisher::handleNodeInfoUnavailable(uavcan::NodeID node_id)
+{
+}
+
+void NodeInfoPublisher::handleTimerEvent(const uavcan::TimerEvent &event)
+{
+	// Publish device information using round-robin approach
+	publishDeviceInformationPeriodic();
+}
+
+void NodeInfoPublisher::startTimerIfNotRunning()
+{
+	if (!TimerBase::isRunning()) {
+		TimerBase::startPeriodic(uavcan::MonotonicDuration::fromMSec(DEVICE_INFO_PUBLISH_INTERVAL_MS));
+	}
+}
+
+void NodeInfoPublisher::registerDevice(uint8_t node_id, const NodeInfo *info, uint32_t device_id,
+				       DeviceCapability capability)
+{
+	const bool is_registering_info = (info != nullptr);
+
+	for (size_t i = 0; i < _device_informations_size; ++i) {
+		if (is_registering_info) {
+			// Case 1: Check if this exact node already has node info - skip
+			if (_device_informations[i].node_id == node_id &&
+			    _device_informations[i].has_node_info) {
+				return;
+			}
+
+			// Case 2: Check if node_id already exists with capability but no info - update that entry
+			if (_device_informations[i].node_id == node_id &&
+			    _device_informations[i].has_capability &&
+			    !_device_informations[i].has_node_info) {
+
+				populateDeviceInfoFields(_device_informations[i], *info);
+				publishDeviceInformationImmediate(i);
+				// Continue to check for other entries with the same node_id but different capabilities.
+			}
+
+		} else { // registering capabilities
+			// Case 1: Check if this exact capability already exists - skip
+			if (_device_informations[i].node_id == node_id &&
+			    _device_informations[i].device_id == device_id &&
+			    _device_informations[i].capability == capability &&
+			    _device_informations[i].has_capability) {
+				return;
+			}
+
+			// Case 1b: if this node has multiple capabilities, add a new entry for the new capability
+			if (_device_informations[i].node_id == node_id &&
+			    _device_informations[i].has_capability &&
+			    _device_informations[i].capability != capability) {
+				if (extendDeviceInformationsArray()) {
```

Comment:
How do you know that you need to extend here? The current capability & node could already exist in `i+1` or later.